### DISCUSSION TITLE
How long to create the minikube cluster

### DISCUSSION THREAD

**Discussion by bdresdner** [eclipse-theia/theia-cloud#248] [Q&A] [answered]
I have followed the steps in Getting started with Minikube to [Create Minikube Cluster](https://github.com/eclipsesource/theia-cloud/blob/main/terraform/terraform.md#create-minikube-cluster).  When running the step, `terraform apply`, I can see the messages about the various objects get created: [terraform_apply_output.txt](https://github.com/eclipsesource/theia-cloud/files/13189551/terraform_apply_output.txt), and then it starts to create the cluster: 
![minikubecluster](https://github.com/eclipsesource/theia-cloud/assets/4948279/2cdab885-c2b2-47db-96f5-7c3c38b79913)
My company grants us 2 hours of Administrator access.  So the process never completes.
My question is: How long should it take to create the cluster?  Could this excessive time be a problem on my end?

**Comment by sgraband** [comment]
Hey @bdresdner,

while the creation of the minikube cluster might take a few minutes it should definately not take 2 hours. Could you try to create a minikube cluster manually and see how long this takes or if it works?

```sh
minikube start --vm=true --driver=virtualbox --download-only
```

**Reply by bdresdner** [reply]
@sgraband Thanks for the reply!

Here is the output:
`PS C:\Development\cloud\theia-cloud\terraform\configurations\minikube_getting_started> minikube start --vm=true --driver=virtualbox --download-only
* minikube v1.31.2 on Microsoft Windows 10 Enterprise 10.0.19045.3570 Build 19045.3570
* Using the virtualbox driver based on existing profile
* Starting control plane node minikube in cluster minikube
* Download complete!`

It happens almost instantly.

**Reply by bdresdner** [reply]
But this isn't the problem.  The problem happens when running `terraform apply`.

**Reply by bdresdner** [reply]
I restarted minikube.  Reran terraform commands: `terraform init`, `terraform plan`, `terraform apply`

Now I get:

`Plan: 13 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  + try_now = (known after apply)

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

module.cluster.minikube_cluster.cluster: Creating...
module.cluster.minikube_cluster.cluster: Still creating... [10s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [20s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [30s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [40s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [50s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [1m0s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [1m10s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [1m20s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [1m30s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [1m40s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [1m50s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [2m0s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [2m10s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [2m20s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [2m30s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [2m40s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [2m50s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [3m0s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [3m10s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [3m20s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [3m30s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [3m40s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [3m50s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [4m0s elapsed]
╷
│ Error: Plugin did not respond
│
│   with module.cluster.minikube_cluster.cluster,
│   on ..\..\modules\cluster_creation\minikube\main.tf line 26, in resource "minikube_cluster" "cluster":
│   26: resource "minikube_cluster" "cluster" {
│
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.`

**Reply by bdresdner** [reply]
Logs:
[minikube_logs.log](https://github.com/eclipsesource/theia-cloud/files/13191736/minikube_logs.log)
[minikube.txt](https://github.com/eclipsesource/theia-cloud/files/13191739/minikube.txt)

**Reply by lucas-koehler** [reply]
Hi @bdresdner , as this seems to be a problem with terraform and/or the minikube provider plugin, have you tried using a later version of the minikube terraform provider? You can update its version here: https://github.com/eclipsesource/theia-cloud/blob/e957794d4d50ffbb2923658c2b0fb362c9953201/terraform/configurations/minikube_getting_started/versions.tf#L5

Looking at https://registry.terraform.io/providers/scott-the-programmer/minikube/latest, the latest version should be `0.3.5`.

**Reply by bdresdner** [reply]
Hi @lucas-koehler Thank you for the response.  I just tried changing as you suggested.  I am now seeing the following when I run `terraform init`:

Initializing the backend...
Initializing modules...

Initializing provider plugins...
- Reusing previous version of scott-the-programmer/minikube from the dependency lock file
- Reusing previous version of hashicorp/external from the dependency lock file
- Reusing previous version of hashicorp/helm from the dependency lock file
- Reusing previous version of hashicorp/kubernetes from the dependency lock file
- Reusing previous version of mrparkers/keycloak from the dependency lock file
- Reusing previous version of gavinbunney/kubectl from the dependency lock file
- Installing mrparkers/keycloak v4.2.0...
- Installed mrparkers/keycloak v4.2.0 (self-signed, key ID C50867915E116CD2)
- Installing gavinbunney/kubectl v1.14.0...
- Installed gavinbunney/kubectl v1.14.0 (self-signed, key ID AD64217B5ADD572F)
- Installing hashicorp/external v2.3.1...
- Installed hashicorp/external v2.3.1 (signed by HashiCorp)
- Installing hashicorp/helm v2.9.0...
- Installed hashicorp/helm v2.9.0 (signed by HashiCorp)
- Installing hashicorp/kubernetes v2.19.0...
- Installed hashicorp/kubernetes v2.19.0 (signed by HashiCorp)

Partner and community providers are signed by their developers.
If you'd like to know more about provider signing, you can read about it here:
https://www.terraform.io/docs/cli/plugins/signing.html
╷
│ Error: Failed to query available provider packages
│
│ Could not retrieve the list of available versions for provider scott-the-programmer/minikube: locked provider registry.terraform.io/scott-the-programmer/minikube 0.2.3 does not match configured version constraint 0.2.3,
│ 0.3.5; must use terraform init -upgrade to allow selection of new versions

Could there be another place I need to modify?  I am new to using terraform.

**Reply by lucas-koehler** [reply]
It seems like it aborts because the new version is not allowed by the lock file pinning the versions.
You can either try running `terraform init -upgrade` as suggested in the output or delete the file `.terraform.lock.hcl` and then run `terraform init` again.

**Reply by bdresdner** [reply]
@lucas-koehler I tried your suggestion, and here is the output:
Initializing the backend...
Initializing modules...
Downloading registry.terraform.io/matti/urlparse/external 0.2.0 for host...
- host in .terraform\modules\host

Initializing provider plugins...
- Finding hashicorp/kubernetes versions matching ">= 2.18.1"...
- Finding mrparkers/keycloak versions matching ">= 4.2.0"...
- Finding gavinbunney/kubectl versions matching ">= 1.14.0"...
- Finding scott-the-programmer/minikube versions matching "0.2.3, 0.3.5"...
- Finding hashicorp/helm versions matching ">= 2.9.0"...
- Finding latest version of hashicorp/external...
- Installing hashicorp/external v2.3.1...
- Installed hashicorp/external v2.3.1 (signed by HashiCorp)
- Installing hashicorp/kubernetes v2.23.0...
- Installed hashicorp/kubernetes v2.23.0 (signed by HashiCorp)
- Installing mrparkers/keycloak v4.3.1...
- Installed mrparkers/keycloak v4.3.1 (self-signed, key ID C50867915E116CD2)
- Installing gavinbunney/kubectl v1.14.0...
- Installed gavinbunney/kubectl v1.14.0 (self-signed, key ID AD64217B5ADD572F)
- Installing hashicorp/helm v2.11.0...
- Installed hashicorp/helm v2.11.0 (signed by HashiCorp)

Partner and community providers are signed by their developers.
If you'd like to know more about provider signing, you can read about it here:
https://www.terraform.io/docs/cli/plugins/signing.html
╷
│ Error: Failed to query available provider packages
│
│ Could not retrieve the list of available versions for provider scott-the-programmer/minikube: no available releases match the given constraints 0.2.3, 0.3.5

Do you know where these constraints are set?

**Reply by lucas-koehler** [reply]
Hi @bdresdner , AFAIK they are only specified in the versions.tf file and then automatically added to the `.terraform.lock.hcl` file. Did you clean the `.terraform` folder after changing the version and deleting the lock file. Maybe some information or constraints are also stored there.

**Reply by bdresdner** [reply]
Hi @lucas-koehler, Just so I am clear:
I deleted my cluster
I deleted theia-cloud/terraform/configurations/minikube_getting_started/.terraform
I also deleted theia-cloud/terraform/configurations/minikube_getting_started/.terraform.lock.hcl
I modified theia-cloud/terraform/configurations/minikube_getting_started/versions.tf (changed line 5 to: `version = "0.3.5"`)
I then ran: `minikube start --vm=true --driver=virtualbox --download-only`
Then I ran: `terraform init`
I get the same error: 
```
│ Error: Failed to query available provider packages
│
│ Could not retrieve the list of available versions for provider scott-the-programmer/minikube: no available releases match the given constraints 0.2.3, 0.3.5
```
Am I doing this correctly?

**Reply by lucas-koehler** [reply]
Hi @bdresdner 
Yes, I think this is correct. However, what tripped me up is that it still searches for a `0.2.3` so I investigated again. It seems that the getting started module is using another module in the repo: The minikube repo at terraform/modules/cluster_creation/minikube. I think you'll need to update the version to `0.3.5` there, too, and possibly also delete the lock file there.

**Reply by bdresdner** [reply]
Thank you @lucas-koehler.  Now I can initialize!  I can also `run terraform` plan successfully.  However, when I run `terraform apply`, I get the following:
```
Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

module.cluster.minikube_cluster.cluster: Creating...
module.cluster.minikube_cluster.cluster: Still creating... [10s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [20s elapsed]
module.cluster.minikube_cluster.cluster: Still creating... [30s elapsed]
╷
│ Error: Failed to start host: creating host: create: precreate: This computer doesn't have VT-X/AMD-v enabled. Enabling it in the BIOS is mandatory
│
│   with module.cluster.minikube_cluster.cluster,
│   on ..\..\modules\cluster_creation\minikube\main.tf line 26, in resource "minikube_cluster" "cluster":
│   26: resource "minikube_cluster" "cluster" {
```
I have Virtualization enabled in my bios, and I can see that it is set when I run Task Managers.  I am running Windows 10.

**Reply by lucas-koehler** [reply, marked as answer]
Hi @bdresdner ,
unfortunately I cannot help you with that as this seems to be specific issue of using terraform, Minikube and Windows. I only have a Ubuntu machine. A quick google search for the error message revealed to me that there is a flag (`--no-vtx-check`) to skip this check: =https://github.com/kubernetes/minikube/issues/3900#issuecomment-474523309

Thus, you could try adding the flag `no_vtx_check      = true` to `terraform/modules/cluster_creation/minikube/main.tf` like this:
![image](https://github.com/eclipsesource/theia-cloud/assets/6959840/cea8e334-25f3-4565-8d15-820df080dba5)


Please also note that we do this support besides our actual work. If you need in depth help covered by an NDA, see [professional support options](https://eclipsesource.com/services/developer-support/).

**Reply by bdresdner** [reply]
Hi @lucas-koehler - Thanks a lot for your help.  That solved the problem, and I am able to create the cluster now!
